In [1]:
import numpy as np

In [2]:
def aryinfo(ary):
    print('type : {}'.format(type(ary)))
    print('shape : {}'.format(ary.shape)) # 차원의 크기를 튜플(tuple) 형태로 나타냄
    print('dimension:{}'.format(ary.ndim))
    print('dtype:{}'.format(ary.dtype))
    print('Array Data:\n',ary)

- 넘파이 array() 함수는 파이썬의 리스트와 같은 다양한 인자를 입력 받아서 ndarray로 변환하는 기능 수행
- ndarry배열의 shape 변수는 ndarray의 크기, 즉 행과 열의 수를 튜플 형태로 가지고 있음

In [4]:
# 1 차원 배열
array1=np.array([1,2,3])
aryinfo(array1)

type : <class 'numpy.ndarray'>
shape : (3,)
dimension:1
dtype:int32
Array Data:
 [1 2 3]


In [8]:
# 2차원 배열
array2=np.array([[1,2,3],[2,3,4]])
aryinfo(array2)
print('--------')
array3=np.array([[1,2,3]]) # 2차원이다.
aryinfo(array3)

type : <class 'numpy.ndarray'>
shape : (2, 3)
dimension:2
dtype:int32
Array Data:
 [[1 2 3]
 [2 3 4]]
--------
type : <class 'numpy.ndarray'>
shape : (1, 3)
dimension:2
dtype:int32
Array Data:
 [[1 2 3]]


In [11]:
squence_array=np.arange(10)
squence_array
aryinfo(squence_array)

type : <class 'numpy.ndarray'>
shape : (10,)
dimension:1
dtype:int32
Array Data:
 [0 1 2 3 4 5 6 7 8 9]


In [17]:
zero_array=np.zeros((3,2))
print(zero_array)
print('-------')
one_array=np.ones((3,2))
print(one_array)

[[0. 0.]
 [0. 0.]
 [0. 0.]]
-------
[[1. 1.]
 [1. 1.]
 [1. 1.]]


In [22]:
array1=np.arange(10)
array2=array1.reshape(2,5)
print(array2)
print('-------')
array3=array1.reshape(5,2)
print(array3)

[[0 1 2 3 4]
 [5 6 7 8 9]]
-------
[[0 1]
 [2 3]
 [4 5]
 [6 7]
 [8 9]]


In [32]:
array1d=np.arange(1,10)
array1d

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [34]:
array2d=array1d.reshape(3,3)
aryinfo(array2d)

type : <class 'numpy.ndarray'>
shape : (3, 3)
dimension:2
dtype:int32
Array Data:
 [[1 2 3]
 [4 5 6]
 [7 8 9]]


In [36]:
array2d[0:2,0:2]

array([[1, 2],
       [4, 5]])

In [37]:
array2d[1:3,:]

array([[4, 5, 6],
       [7, 8, 9]])

In [38]:
array2d[:,:]

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

In [40]:
array2d[:2,1:]

array([[2, 3],
       [5, 6]])

In [41]:
array2d[:2,0]

array([1, 4])

##### 펜싱 인덱싱

- 리스트나 ndarry로 인덱스 집합을 지정하면 해당 위치의 인덱스에 해당하는 ndarray를 반환하는 인덱싱 방식

In [44]:
array2d[[0,1],2]

array([3, 6])

In [46]:
array2d[[0,1],0:2]

array([[1, 2],
       [4, 5]])

In [48]:
array2d[[0,1]]

array([[1, 2, 3],
       [4, 5, 6]])

##### 불린 인덱싱
- 조건 필터링과 검색을 동시에 할 수 있기 때문에 매우 자주 사용되는 인덱싱이다. 

In [50]:
array1d=np.arange(1,10)
array1d

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [52]:
array3=array1d[array1d>5]
array3

array([6, 7, 8, 9])

#####  행렬의 정렬   =  sort(),   argsort()
- 넘파이에서 행렬을 정렬하는 대표적인 방법인 np.sort() 와 ndarray.sort() 그리고 정렬된 행렬의 인덱스를 반환하는 argsort()가 있다

In [55]:
org_array=np.array([3,1,9,5])
print('원본행렬 : ',org_array)

원본행렬 :  [3 1 9 5]


In [56]:
sort_array1=np.sort(org_array)
print('np.sort() 호출 후 반환된 정렬 행렬: ',sort_array1)

np.sort() 호출 후 반환된 정렬 행렬:  [1 3 5 9]


In [60]:
sort_array2=org_array.sort()
print('org_array.sort() 호출 후 반환된 행렬: ',sort_array2)
print('org_array.sort() 호출 후 반환된 원본 행렬: ',org_array)

org_array.sort() 호출 후 반환된 행렬:  None
org_array.sort() 호출 후 반환된 원본 행렬:  [1 3 5 9]


In [64]:
org_array=np.array([3,1,9,5])
sort_indices=np.argsort(org_array)
print(type(sort_indices))
print("행렬 정렬시 원본 행렬의 인덱스 :",sort_indices)

<class 'numpy.ndarray'>
행렬 정렬시 원본 행렬의 인덱스 : [1 0 3 2]


In [65]:
org_array=np.array([3,1,9,5])
sort_indices_desc=np.argsort(org_array)[::-1]
print(type(sort_indices))
print("행렬 내림차순 정렬시 원본 행렬의 인덱스 :",sort_indices_desc)

<class 'numpy.ndarray'>
행렬 내림차순 정렬시 원본 행렬의 인덱스 : [2 3 0 1]


#### 선형대수 연산 - 행렬 내적과 전치 행렬 구하기

- 행렬 내적(행렬 곱)   -> np.dot()

In [67]:
A=np.array([[1,2,3],[4,5,6]])
B=np.array([[7,8],[9,10],[11,12]])

In [68]:
np.dot(A,B)

array([[ 58,  64],
       [139, 154]])

- 전치 행렬 : 원 행렬에서 행과 열 위치를 교환한 원소를 구성한 행렬을 그 행렬의 전치 행렬이라고 한다.

In [72]:
A=np.array([[1,2],[3,4]])
transpose_mat=np.transpose(A)
print('A의 전치행렬 : \n', transpose_mat)

A의 전치행렬 : 
 [[1 3]
 [2 4]]


### 데이터 핸들링 - 판다스

In [73]:
import pandas as pd

In [86]:
titanic_df=pd.read_csv(r"C:\Users\mgk04.DESKTOP-8Q51H2D\Desktop\train.csv")
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [87]:
print('DataFrame 크기 :', titanic_df.shape) #DataFrame의 행과 열을 튜플 형태로 반환
# DataFrame은 데이터뿐만 아니라 칼럼의 타입, Null 데이터 개수, 데이터 분포도 등의 메타 데이터 등도 조회 가능
# 대표적으로 info()와  describe()가 있음.

DataFrame 크기 : (891, 12)


In [89]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 83.6+ KB


##### info() 메서드를 통해서 총 데이터 건수와 데이터 타입, Null건수를 알수 있다.
- RangeIndex는 DataFrame index의 범위의 나타냄 전체 데이터 891개 row이다.
- 몇 개의 데이터가 non-null(Null 값이 아님)인지 나타냅니다.
- dtypes은 전체 12개의 칼럼들의 타입을 요약한 것임

In [91]:
titanic_df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [92]:
values_counts=titanic_df['Pclass'].value_counts()
values_counts #value_countss() 메서드는 Series 객체에서만 정의. DataFrame에서 X

3    491
1    216
2    184
Name: Pclass, dtype: int64

In [98]:
import numpy as np


col_name1=['col']
list1=[1,2,3]
array1=np.array(list1)
print('array1 shape :',array1.shape)
# 1차원 형태의 데이터를 기반으로 DataFrame을 생성하므로 컬럼명이 한개여야함 (주의) 

# 리스트를 이용해 DataFrame 생성.
df_list1=pd.DataFrame(list1,columns=col_name1)
print('1차원 리스트로 만든 DataFrame: ' ,df_list1)
print('-------------')
# 넘파이 ndarray를 이용해 DataFrame 생성
df_array1=pd.DataFrame(array1,columns=col_name1)
print('1차원 ndarray로 만든 DataFrame: ',df_array1)

array1 shape : (3,)
1차원 리스트로 만든 DataFrame:     col
0    1
1    2
2    3
-------------
1차원 ndarray로 만든 DataFrame:     col
0    1
1    2
2    3


In [102]:
# 2차원 형태의 데이터를 기반으로 DataFrame을 생성
# 2행 3열 형태의 리스트와 ndarray를 기반으로  DataFrame을 생성하므로 칼럼명은 3개가 필요함

# 3개의 칼럼명이 필요함
col_name2=['col1','col2','col3']

list2=[[1,2,3],[11,12,13]]
array2=np.array(list2)
print('array2 shape :',array2.shape)
df_list2=pd.DataFrame(list2,columns=col_name2)
print('2차원 리스트로 만든 DataFrame:\n ' ,df_list2)
df_array2=pd.DataFrame(array2,columns=col_name2)
print('2차원 ndarray로 만든 DataFrame:\n ',df_array2)

array2 shape : (2, 3)
2차원 리스트로 만든 DataFrame:
     col1  col2  col3
0     1     2     3
1    11    12    13
2차원 ndarray로 만든 DataFrame:
     col1  col2  col3
0     1     2     3
1    11    12    13


In [109]:
# Key는 문자열 칼럼명으로 매핑, Value는 리스트형 칼럼 데이터로 매핑
dict={'col1':[1,11],'col2':[2,22],'col3':[3,33]}
df_dict=pd.DataFrame(dict)
print('딕셔너리로 만든 DataFrame: \n',df_dict)

딕셔너리로 만든 DataFrame: 
    col1  col2  col3
0     1     2     3
1    11    22    33


##### DataFrame을 넘파이 ndarray, 리스트, 딕셔너리로 변환하기

- values를 이용한 ndarray의 변환은 매우 많이 사용된다

In [110]:
# DataFrame을 ndarray로 변환
array3=df_dict.values
print('df_dict.values타입 :', type(array3))
print(array3)

df_dict.values타입 : <class 'numpy.ndarray'>
[[ 1  2  3]
 [11 22 33]]


##### DataFrame을 리스트와 딕셔너리로 변환


- 리스트로 변환은 values로 얻은 ndarray에 tolist()를 호출
- 딕셔너리 변환은 DataFrame 객체의 to_dict() 메서드를 호출하는데 인자로 'list'를 입력하면 딕셔너리의 값이 리스트형으로 반환

In [114]:
# DataFrame을 리스트로 변환
list3=df_dict.values.tolist()
print(list3)

[[1, 2, 3], [11, 22, 33]]


In [117]:
# DataFrame을 딕셔너리로 벼환
dict3=df_dict.to_dict()
print(dict3)

{'col1': {0: 1, 1: 11}, 'col2': {0: 2, 1: 22}, 'col3': {0: 3, 1: 33}}


##### DataFrame의 칼럼 데이터 세트 생성과 수정

In [118]:
titanic_df['Age_0']=0
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  
0      0         A/5 21171   7.2500   NaN        S      0  
1      0          PC 17599  71.2833   C85        C      0  
2      0  STON/O2. 3101282   7.9250   NaN        S      0

In [119]:
titanic_df['Age_by_10']=titanic_df['Age']*10
titanic_df['Family_No']=titanic_df['SibSp']+titanic_df['Parch']
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      220.0   
1      0          PC 17599  71.2833   C85        C      0      380.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      260.0   

   Family_No  
0          1  
1          1  
2          0

In [120]:
# 업데이트도 가능
titanic_df['Age_by_10']=titanic_df['Age_by_10']+100
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          1  
1          1  
2          0

##### DataFrame 데이터 삭제

- DataFrame에서 데이터 삭제는 drop() 메서드 이용

- DataFrame.drop(labels=None,axis=0,index=None,cloumns=None,level=None,inplace=None,errors='raise')

- axis에 따라서 특정 칼럼 또는 특정 행을 드롭한다. axis 0은 로우 방향 축 , axis 1은 칼럼 방향 축
- labels에 원하는 칼럼 명을 입력하고 axis=1을 입력하면 지정된 칼럼을 드롭
- axis=0으로 설정하고 로우 레벨로 삭제하는 경우는 이상치 데이터를 삭제하는 경우에 주로 사용

In [125]:
titanic_drop_df=titanic_df.drop('Age_0',axis=1)
print(titanic_drop_df.head(3))

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_by_10  Family_No  
0      0         A/5 21171   7.2500   NaN        S      320.0          1  
1      0          PC 17599  71.2833   C85        C      480.0          1  
2      0  STON/O2. 3101282   7.9250   NaN        S      360.0          0  


In [126]:
titanic_drop_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_by_10  Family_No  
0      0         A/5 21171   7.2500   NaN        S      320.0          1  
1      0          PC 17599  71.2833   C85        C      480.0          1  
2      0  STON/O2. 3101282   7.9250   NaN        S      360.0          0

##### index 객체

In [131]:
titanic_df=pd.read_csv(r"C:\Users\mgk04.DESKTOP-8Q51H2D\Desktop\train.csv")
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [133]:
# index 객체 추출
indexes=titanic_df.index
print(indexes)
# index 객체를 실제 값 array로 변환
print('index객체 array값:\n',indexes.values)

RangeIndex(start=0, stop=891, step=1)
index객체 array값:
 [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 

- index객체는 식별성 데이터를 1차원 array로 가지고 있다. 또한 ndarray와 유사하게 단일 값 반환 및 슬라이싱도 가능하다 

In [134]:
print(type(indexes.values))
print(indexes.values.shape)

<class 'numpy.ndarray'>
(891,)


- 한번만들어진 DataFrame 및 series의 index객체는 함부로 변경할 수 없다. 
- 즉 다음과 같이 첫번째 index 객체의 값을 5로 변경하는 작업은 수행 X